driver/lchmod_linux.go

```diff
@@ -0,0 +1,19 @@
+package driver
+
+import (
+	"os"
+
+	"github.com/containerd/continuity/sysx"
+)
+
+// Lchmod changes the mode of a file not following symlinks.
+func (d *driver) Lchmod(path string, mode os.FileMode) error {
+	// On Linux, file mode is not supported for symlinks,
+	// and fchmodat() does not support AT_SYMLINK_NOFOLLOW,
+	// so symlinks need to be skipped entirely.
+	if st, err := os.Stat(path); err == nil && st.Mode()&os.ModeSymlink != 0 {
```

Comment:
(Perhaps overthinking it, but leaving this here because I learned to never "assume" things 😅)

From https://go-review.googlesource.com/c/go/+/118658

```
Currently Linux' fchmodat(2) syscall implementation doesn't support the
flags parameter (though it might in future versions [1]). Fchmodat in
the syscall package takes the parameter and (wrongly) passes it on to the
syscall which will ignore it.

According to the POSIX.1-2008 manual page [2], AT_SYMLINK_NOFOLLOW is
the only valid value for the flags parameter and EOPNOTSUPP should be
returned in case changing the mode of a symbolic link is not supported
by the underlying system. EINVAL should be returned for any other value
of the flags parameter.

  [1] https://patchwork.kernel.org/patch/9596301/
  [2] http://pubs.opengroup.org/onlinepubs/9699919799/functions/chmod.html
```

That patch mentions the possibility that in future, other options may actually
become active.

Instead of _skipping_ `sysx.Fchmodat()` for symlinks, should we always run it, but check if the returned error was an `EOPNOTSUPP`, _then_ return `nil` if the path was a symlink?

Reply:
Well, one problem is `syscall.Fchmodat()` from golang < 1.11 does not return the error (same for unix.Fchmodat() from golang.org/x/sys/unix before https://github.com/golang/sys/commit/c23410a886927bab8ca5e80b08af6a56faeb330d which only landed about a year ago), instead changing the mode for the file that a symlink points to (bad bad bad!). So we can't rely on it returning the proper error, and from the correctness/security standpoint it makes sense to avoid the call.

Other things is, Linux won't support mode for symlinks, ever (I mean, I can't speak for Linus, but it looks like it). The patch being referred to above (https://patchwork.kernel.org/patch/9596301/) only adds flags argument to fchmodat(), for the sake of returning an error in case flags are specified. Even that one didn't made it to the kernel, and it seems to be the 3rd attempt, previous two happened in 2015 (https://groups.google.com/d/msg/linux.kernel/Z2z3_MR2aM8/mFz04xQhZEIJ) and 2014 (https://groups.google.com/d/msg/linux.kernel/8lnkJ-naNRI/QvkfwNih6rcJ).

Reply:
Please ignore the previous comment; while correct it misses the biggest issue, that is, EOPNOTSUPP is returned _always_, not just for the symlinks. In other words, calling `Fchmodat()` with flags != 0 is useless, it does nothing but returns an error, and it happens for all files, not just symlinks.

Reply:
Thanks for doing the digging; looks like the current approach is best then!